kfctl updates for v0.7 - GCP docs only

sarahmaddox opened the pull request:
This PR updates the docs in the GCP section to use the new kfctl semantics in v0.7.

The PR addresses some of the updates for https://github.com/kubeflow/website/issues/1226. Further PRs will update non-GCP docs.

This PR also addresses the config URL changes for GCP config files, thus partially addressing https://github.com/kubeflow/website/issues/1224

This PR also reduces the confusion around deployment name and directory, thus partially addressing issue https://github.com/kubeflow/website/issues/1252.
<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/website/1234)
<!-- Reviewable:end -->

sarahmaddox commented:
Preview:
* https://deploy-preview-1234--competent-brattain-de2d6d.netlify.com/docs/gke/deploy/deploy-cli/
* https://deploy-preview-1234--competent-brattain-de2d6d.netlify.com/docs/gke/customizing-gke/
* https://deploy-preview-1234--competent-brattain-de2d6d.netlify.com/docs/gke/deploy/monitor-iap-setup/
* https://deploy-preview-1234--competent-brattain-de2d6d.netlify.com/docs/gke/deploy/delete-cli/


**Note:** I've set the config URLs to interim paths in `/kubeflow/manifests/master/`. ([Example](https://raw.githubusercontent.com/kubeflow/manifests/master/kfdef/kfctl_gcp_iap.yaml).) When we know the path for the v0.7 RC1 release, we must update these paths.

sarahmaddox commented:
This PR is now ready for review. I'm applying a hold to allow for multiple reviewers.
/hold

sarahmaddox commented:
/assign @jlewi 
/assign @swiftdiaries

sarahmaddox commented:
/assign @thedriftofwords

jlewi commented on content/docs/gke/customizing-gke.md:
````diff
@@ -7,56 +7,104 @@ weight = 2
 
-You can use [kustomize](https://kustomize.io/) to customize Kubeflow.
+  ```
+  export KFAPP=<your choice of application directory name>
````
nit: I would say name of Kubeflow deployment; The directory is one "resource" that gets named after the deployment.

sarahmaddox replied:
Done.

jlewi commented on content/docs/gke/customizing-gke.md:
````diff
@@ -7,56 +7,104 @@ weight = 2
+  `${KFAPP}` directory:
+
+  ```
+  cd <the parent directory of your KFAPP directory>
````
I think people have gotten confused about absolute directory names vs. app names etc...

Would it help if we wrote this way

```
export APPSDIR=<Directory where you'd like your Kubeflow deployments to be stored>
export KFNAME=<Name for your kubeflow deployment>
export KFAPPDIR=${APPSDIR}/${KFNAME}

mkdir -p ${KFAPPDIR}
cd ${KFDAPPDIR}
```

sarahmaddox replied:
Yes, that's a good idea. I'll hive that off to a separate issue, because it means wide-spread updates throughout the docs. Many docs rely on the semantics of `${KFAPP)`. I've created this issue: https://github.com/kubeflow/website/issues/1252

Let's do this outside the scope of a particular release - so, when the docs for Kubeflow v0.7 are safely bedded down.

jlewi commented on content/docs/gke/customizing-gke.md:
```diff
@@ -7,56 +7,104 @@ weight = 2
-  * **DEPLOYMENT_NAME** Name of your Kubeflow app. You could also find it in `${KFAPP}/app.yaml`. 
-    In specific, `.metadata.name`
+  * **PROJECT** The ID of your GCP project. You can find the project ID in 
+    your `${KFAPP}/app.yaml` file, as the value for the `project` key.
```
${KFAPP}/app.yaml should change hopefully by 0.7.0

It should be

${KFAPPDIR}/${file_name}

Where filename is just the filename of the file they applied  e.g.

if they do

kfctl apply -f https://github.com/kfctl_gcp_iap.yaml

Then the command here would be

kfctl apply -f kfctl_gcp_iap.yaml

Or they could specify the full path.

kfctl apply -f ${KFAPPDIR}/kfctl_gcp_iap.yaml

Another way to think of it and possibly explain it is that users should always do

kfctl apply -f ${KFAPPDIR}/${KF_CONFIG_FILE}

The "one exception" is if the file is a remote URI e.g. 
kfctl apply -f https://....

In this case when a user runs kfctl apply -f we will copy the remote file to ${KFAPPDIR}/

All subsequent commands should then be kfctl apply -f ${KFAPPDIR}/${KFFILE}

sarahmaddox replied:
Thanks @jlewi - it seems this change is not certain for v0.7? If it's not certain, then let's put this change into a separate issue, which we can address when the change happens. LMK and I'm happy to create the issue.

jlewi replied:
I think it needs to land before we finalize the RC. Its a P0 bug.
https://github.com/kubeflow/kfctl/issues/54

jlewi replied:
I feel more comfortable with updating the docs to reflect the desired syntax. From a 0.7 perspective if the file gets renamed to "app.yaml" that's a P0 bug. So I'd rather have the docs be consistent with that expectation as opposed to risk setting an expectation that "app.yaml" is working as intended and then we are changing the behavior.

sarahmaddox replied:
Sounds good. Done.

jlewi commented on content/docs/gke/deploy/deploy-cli.md:
```diff
@@ -32,92 +32,130 @@ Before installing Kubeflow on the command line:
+  [Compute Engine documentation](https://cloud.google.com/compute/docs/regions-zones/#available).
+  If you plan to use accelerators, make sure to pick a zone that supports the 
+  type you want.
+* `kfctl apply` attempts to fetch your email address from your 
```
from gcloud; i.e User can run gcloud auth list to see what the default account is.

sarahmaddox replied:
Nice. Done.

jlewi commented on content/docs/gke/deploy/deploy-cli.md:
```diff
@@ -32,92 +32,130 @@ Before installing Kubeflow on the command line:
+
+Notes:
+
+* **${KFAPP}** - The _name_ of a directory where you want Kubeflow 
```
Name, Project, Zone, and Email can also be explicitly set by setting them in the config file.

If any of those values aren't set kfctl will attempt to get defaults as follows
gcloud for project, zone, and email (users can do gcloud config list to see the values).

NAME will be inferred from the directory where kfctl is run.

The following is a snippet that may be useful for helping people understand how to set them explicitly.

```
yq w -i kfctl_gcp_iap.yaml spec.plugins[0].spec.project ${PROJECT}
yq w -i kfctl_gcp_iap.yaml spec.plugins[0].spec.zone ${ZONE}
yq w -i kfctl_gcp_iap.yaml metadata.name ${KFNAME}
```

This is using yq from here.
https://github.com/mikefarah/yq/releases


* Project, Zone, and Email from the gcloud

sarahmaddox replied:
Thanks. Added.

jlewi commented on content/docs/gke/deploy/deploy-cli.md:
````diff
@@ -132,72 +170,77 @@ Follow these steps to deploy Kubeflow:
+  minutes for the URI to become available:
 
-## Understanding the deployment process
+  ```
````
Users can also run the following command to get the host

```
kubectl -n istio-system get ingress
NAME            HOSTS                                          ADDRESS         PORTS   AGE
envoy-ingress   jlewi-v07-002.endpoints.jlewi-dev.cloud.goog   34.102.232.34   80      5d13h
```

Here's a one liner to set an environment variable to the host

```
export HOST=$(kubectl -n istio-system get ingress envoy-ingress -o=jsonpath={.spec.rules[0].host})
```

sarahmaddox replied:
Nice. Added.

jlewi commented on content/docs/gke/deploy/deploy-cli.md:
```diff
@@ -132,72 +170,77 @@ Follow these steps to deploy Kubeflow:
+Your Kubeflow app directory **${KFAPP}** contains the following files and 
+directories:
 
 * **app.yaml** defines configurations related to your Kubeflow deployment.
```
The YAML file will be whatever the user supplied to -f

sarahmaddox replied:
I'll address this when we know the answer to the previous comment, about whether this change is definitely in v0.7.

jlewi replied:
The change will definitely be in 0.7; until it is 0.7 is broken.

sarahmaddox replied:
Sounds good. Done. We must make sure we don't publish the updated docs until this change is in, assuming we publish the docs for one of the RCs.

jlewi commented on layouts/shortcodes/config-uri-gcp-basic-auth.html:
```diff
@@ -1 +1 @@
-https://raw.githubusercontent.com/kubeflow/kubeflow/cfb336b7/bootstrap/config/kfctl_gcp_basic_auth.0.6.2.yaml
\ No newline at end of file
+https://raw.githubusercontent.com/kubeflow/manifests/master/kfdef/kfctl_gcp_basic_auth.yaml
```
Change master to v0.7-branch e.g
https://raw.githubusercontent.com/kubeflow/manifests/v0.7-branch/kfdef/kfctl_gcp_iap.yaml

jlewi commented:
Thank you so much @sarahmaddox

sarahmaddox commented:
Thanks @jlewi PTAL

I'll work on getting the other kfctl PRs to the same point as this one, and then discuss with the community whether we're ready to brand the website as v0.7. I think we should be good to do that, once all the deployment guides are correct for the latest v0.7 RC.

If agreed, then I'll merge all the deployment guides and rebrand the site at the same time.

sarahmaddox commented:
Status: I'm busy updating the files for the coming change in name of config file and in 'kfctl delete'. Stay tuned. Thanks @jlewi for the info!

sarahmaddox commented:
@jlewi I've applied all your feedback. PTAL.

sarahmaddox commented:
Thanks @jlewi! The commenting format on this PR has turned weird, so I've copied your comments and replied in line: 

> Users currently have to do one of the following
> 
> 1. Pick a zone with K80s
>    or
> 2. Disable node-autoprovisioning or change node-autoprovisioning config.
> 
> To see which accelerators are in which zones they can run
> 
> ```
>  gcloud compute accelerator-types list
> ```
> 
> To disable node-autoprovisioning they would run
> kfctl build -f ...
> Then edit gcp_config/cluster-kubeflow.yaml and set enabled to false
> https://github.com/kubeflow/manifests/blob/4d2939d6c1a5fd862610382fde130cad33bfef75/gcp/deployment_manager_configs/cluster-kubeflow.yaml#L73

Good info. I've added the info to the [customization guide](https://deploy-preview-1234--competent-brattain-de2d6d.netlify.com/docs/gke/customizing-gke/#gpu-config), and added a link from the deployment guide instead of complicating the deployment guide.

> nit: "This also becomes" -> "This will also be" or "This will be used as the"
Users create ${KF_DIR} below "becomes" implies to me that some other process is creating the directory

Good point. Updated.

> I think the final names will be
https://raw.githubusercontent.com/kubeflow/manifests/v0.7-branch/kfdef/kfctl_gcp_iap.0.7.0.yaml
> Those files don't exist yet but I think you could update the docs to use that pattern so the docs don't need to be updated again once we create those files.

I think I'll keep going with the RC URLs, for two reasons. First, we may decide to publish the docs when we have a good RC. And second, reviewers in the other CLs are sure to point out the broken link. :)

jlewi commented on layouts/shortcodes/kf-latest-version.html:
```diff
@@ -1 +1 @@
-v0.6.2
\ No newline at end of file
+v0.7.0-rc.4
```
Latest version is now rc.5

sarahmaddox replied:
Done.

jlewi commented on layouts/shortcodes/config-uri-gcp-iap.html:
```diff
@@ -1 +1 @@
-https://raw.githubusercontent.com/kubeflow/kubeflow/cfb336b7/bootstrap/config/kfctl_gcp_iap.0.6.2.yaml
\ No newline at end of file
+https://raw.githubusercontent.com/kubeflow/manifests/v0.7-branch/kfdef/kfctl_gcp_iap.yaml
```
This is now kfctl_gcp_iap.0.7.0.yaml

sarahmaddox replied:
Done.

jlewi commented on layouts/shortcodes/config-uri-gcp-basic-auth.html:
```diff
@@ -1 +1 @@
-https://raw.githubusercontent.com/kubeflow/kubeflow/cfb336b7/bootstrap/config/kfctl_gcp_basic_auth.0.6.2.yaml
\ No newline at end of file
+https://raw.githubusercontent.com/kubeflow/manifests/v0.7-branch/kfdef/kfctl_gcp_basic_auth.yaml
```
This is now kfctl_gcp_basic_auth.0.7.0.yaml

sarahmaddox replied:
Done.

jlewi commented:
@sarahmaddox apologies for the delay

/lgtm
/approve

The 0.7.0.yaml files now exist for the GCP configs
https://github.com/kubeflow/manifests/tree/v0.7-branch/kfdef

Did you want to update the links in this PR or get it merged and then we'd fix in a follow on PR.

/hold

sarahmaddox commented:
> 
> The 0.7.0.yaml files now exist for the GCP configs
> https://github.com/kubeflow/manifests/tree/v0.7-branch/kfdef
> 
> Did you want to update the links in this PR or get it merged and then we'd fix in a follow on PR.
> 

Thanks @jlewi! I'll update the version number and YAML URLs in this PR and all related PRs, so that we can submit them at approx the same time. I'll do that tomorrow morning PT.

I'll also put together a PR to brand the website as v0.7.

Are you happy enough with RC5 for us to follow the above steps? (I think publishing the docs and rebranding the website is kind of a statement that we have a good RC.)

jlewi commented:
@sarahmaddox a couple clarifications

* I think only the GCP configs have 0.7.0 suffixes right now; 

  * For AWS and existing (and GCP) we should match what is current in https://github.com/kubeflow/manifests/tree/v0.7-branch/kfdef and if it changes we will update the docs.

* When we are happy with the RC we will remove the RC label from it an publish it just as 0.7.0
   
   * I think it makes sense to publish the docs with reference to an RC release (whichever is the latest) so that we can more people to try things out. 
   * Lets keep the default docs as 0.6 until we final the release (i.e. remove the RC tag)
   * The current RC is looking pretty good but there are still a couple known issues.

sarahmaddox commented:
Thanks @jlewi This PR now reflects the latest RC and config URLs. If you're happy to LGTM, we can merge the PR.

/hold cancel

jlewi commented:
/lgtm
/approve
